### DISCUSSION TITLE
Embedding large cab files in a msi

### DISCUSSION THREAD

**Discussion by anjannath** [wixtoolset/issues#6400] [Questions] [answered]
Hi, I am trying to build an msi for a go binary and it dependents on a large image file that is around 2.3GB, i have split the file into three part files, when i try to build the msi with the cabinet files embedded, all the wixtoolset steps i use (candle; light) goes fine, and i get an msi, upon installing though it fails with the following error.

// screenshots of the error popups
![Untitled](https://user-images.githubusercontent.com/8885742/112495036-8842ca00-8da9-11eb-881c-d34f1ac983bf.png)

// the product.wxs fragment
```xml
<MediaTemplate EmbedCab="yes" />
<Directory Id="TARGETDIR" Name="SourceDir">
      <Directory Id="$(var.Program_Files)">
	   <Directory Id="INSTALLDIR" Name="test">
                   <Component Id="ImageFile1" Guid="*">
                        <File Id="ImageFile1" Source="test.img.0" KeyPath="yes" Checksum="yes" />
                    </Component>
                    <Component Id="ImageFile2" Guid="*">
                        <File Id="ImageFile2" Source="test.img.1" KeyPath="yes" Checksum="yes" />
                    </Component>
                    <Component Id="ImageFile3" Guid="*">
                        <File Id="ImageFile3" Source="test.img.2" KeyPath="yes" Checksum="yes" />
                    </Component>
                    <Component Id="mainExecutable" Guid="*">
                        <File Id="mainExecutable" Source="foo-bar.exe" KeyPath="yes" Checksum="yes" />
                    </Component>
                    <Component Id="AddToPath" Guid="09CXXX3-44DE-4XXX-BDAD-72BE10C10542">
                        <CreateFolder/>
                        <Environment Id="PATH" Name="PATH" Value="[INSTALLDIR]" Permanent="no" Part="last" Action="set" />
                    </Component>
		</Directory>
	</Directory>
</Directory>
```
---------------------------------------------------------------
I also tried to use the `<Media>` element, with premade cabs using `makecab.exe`

I got an error from `light.exe` when i tried to embed the file directly without splitting which i assumed meant for a single cab file, but is there also a limit to make a msi package with embedded cabinet files that add up to more then 2GB although they each are < 2GB?

looks similar to this olde issue: https://github.com/wixtoolset/issues/issues/1155

**Comment by BMurri** [comment]
Look in the MSI log for the lines surrounding "2755". What do they say?

**Reply by anjannath** [reply]
Not much info in the log file, 

```
The system cannot open the device or file specified.
The system cannot open the device or file specified.
The installer has encountered an unexpected error installing this package. This may indicate a problem with this package. The error code is 2755. The arguments are: 110, C:\Users\Anjan\Desktop\msi-package\foo-bar.msi, 
=== Logging stopped: 25/03/2021  20:29:32 ===
```

**Reply by BMurri** [reply]
That [error fully formatted](https://docs.microsoft.com/en-us/windows/win32/msi/windows-installer-error-messages) would be `Server returned unexpected error 110 attempting to install package C:\Users\Anjan\Desktop\msi-package\foo-bar.msi.`. [Error 110](https://docs.microsoft.com/en-us/windows/win32/debug/system-error-codes--0-499-) is

> ERROR_OPEN_FAILED
> 110 (0x6E)
> The system cannot open the device or file specified.

Your next step might be using [Procmon](https://docs.microsoft.com/en-us/sysinternals/downloads/procmon) to see what path was not available to `msiexec.exe` (that process will likely be running as the `Windows Installer` service, so you'll need `procmon` elevated).

**Reply by anjannath** [reply]
Hi @BMurri could not find any useful information using Procmon, the only not SUCCESS messages i saw that originated from `msiexec.exe` were about loading some kernel dlls, it errored with `LOCK ONLY READER`

What i basically wanted to know is, Is there a limit to the size of embedded stuff we can put in the msi? As when i cross 2GB only then these issues come up

**Reply by rseanhall** [reply]
Have you tried to not embed the cabs (have them sit next to the MSI instead)? It sounds like Windows Installer doesn't support an MSI file bigger than 2GB.

**Reply by anjannath** [reply]
@rseanhall Yeah, that's what i also found on the internet but no official sources mentions this size limit (at least i haven't found), and yes by not embedding them it works like charm! But the convenience of having to just ship a single binary is what makes me to use the embedded approach.

**Comment by wyrdfish** [comment]
are you sure you want Checksum on your image fragments? That is for files with a PE header.

**Reply by anjannath** [reply]
Ahh, didn't know that, will be removed then, thanks!

**Comment by rseanhall** [comment, marked as answer]
An MSI file is in the [COM Structured Storage format](https://robmensching.com/blog/posts/2003/11/25/inside-the-msi-file-format/). With the default sector size, that format has a [limit of 2GB](https://docs.microsoft.com/en-us/windows/win32/stg/structured-storage-interfaces). Since the MSI APIs don't expose the sector size when creating a new MSI, it's probably safe to assume that it only supports the default sector size.

You can use a WiX bundle to deliver the MSI with multiple cabs in a single binary.
